File: Memoization-climbingStairs.cpp
70. Climbing Stairs


class Solution {
public:
    int climbStairs(int n) {
        vector<int>dp(n+1);
        
        dp[0]=1;
        dp[1]=1;
        
        for(int i=2;i<=n;i++){
            
            dp[i]=dp[i-1]+dp[i-2];
        }
         return dp[n];
    
    }
};

// memoization
class Solution {
public:
          int solve(vector<int>&dp,int n){
              
              
              if(dp[n]!=-1){
                  return dp[n];
              }
              
          if(n==1 or n==0){
              dp[n]=1;
            return 1;
              }
               
        int ans=solve(dp,n-1)+solve(dp,n-2);
              
              dp[n]=ans;
              
              return ans;
        }
    int climbStairs(int n) {
          
          vector<int>dp(46,-1);
        
         return  solve(dp,n);
    
    }
};

File: Memoization-climbingStairs.py
from typing import Dict

class Solution:
    def memo(self,n: int, dp: Dict[int,int])-> int:
        if n in dp :
            return dp[n]
        if n == 0 :
            return 1 
        if n < 0:
            return 0

        op1 = self.memo(n-1,dp)
        op2 = self.memo(n-2,dp)
        dp[n] = op1 + op2

        return  dp[n]

    def climbStairs(self, n: int) -> int:
        dp = {}
        return self.memo(n,dp)

    
        
